Add support for SAJ H1 hybrid inverters

h3llrais3r opened the pull request:
This PR contains the initial setup for a SAJ H1 inverter and is still work in progress for some additional meter sensors.

H1 inverter is quite similar as the H2 inverter, so I based myself the template for the H2 (https://github.com/evcc-io/evcc/blob/master/templates/definition/meter/saj-h2.yaml)

Documentation about H1: 
https://img.saj-electric.com/file/H1-3-6K-S2-15%C2%A0user%20manual%20EN-V0.0-2023041004191770.pdf
https://github.com/sgsancho/saj_h1_s2_modbus_esp32/blob/main/documentacion/SAJ_Modbus_Communication_Protocol_2020.pdf

Included meters:
- [x] pv
    - [x] power
    - [x] energy 
- [x] battery
    - [x] power
    - [x] energy
    - [x] soc 
    - [x] batterymode
- [x] grid
    - [x] power (smart meter grid power)
    - [ ] ~~currents~~ -> N/A

~~About the grid meter:~~
~~- I found a register which gives the total grid power, so adding the "power" shouldn't be a problem (individual powers for every phase are not available according to my knowledge)~~
~~- For the currents, I only find 1 register which seems to contain the sum of all currents across all phases, not for every single phase. Is it possible to add this single current or not? I only see "currents", not "current" as an option?~~
Grid has been added (power and energy).
Be aware that the grid meter requires the smart meter device (in my case dtsu666) to be installed together with the inverter.

I'm working on the battery control (I already found the similar registers as for the H2) but I would like to understand first how this works. Is there someone who can provide some info about the battery control of the H2 inverter?
As far as I understood the "backup mode" is used by default instead of the "self-use" mode?
@UDicke, it would be great if you could provide me some info as it seems that you did all the testing for the H2 inverter in https://github.com/evcc-io/evcc/pull/15988

h3llrais3r committed:
Add initial template for "pv" and "battery" for SAJ H1 hybrid inverters

andig commented:
> For the currents, I only find 1 register which seems to contain the sum of all currents across all phases, not for every single phase. Is it possible to add this single current or not? I only see "currents", not "current" as an option?

No, but total power is fine.

h3llrais3r commented:
I think I have to omit the "grid" part, as it's not consistent.
I have a 1-phase H1 inverter, added on my 3-phase home installation, so the "grid" sensors are not correct.
I only have correct values from the additional smart meter that can be connected to the inverter.
Since not everyone has this, I think it's better to omit the "grid" part.

premultiply commented:
> I only have correct values from the additional smart meter that can be connected to the inverter.

This is exactly what is needed for grid meter.

andig commented:
1p cannot be a grid meter, so 3p should be the case to add here imho.

h3llrais3r commented:
Ok, I'm going to leave out the grid, as it's too different (I just remember that with an older firmware, the meter readout was even on onther register).
Grid sensor will have to be added manually for this inverter, unless someone can confirm that the following registers are working as well...

```yaml
meters:
  - name: saj.h1.grid
    type: custom
    power:
      source: modbus
      id: 1
      uri: 192.168.0.20:502
      register:
        address: 0x40A7 # TotalGridPowerWatt
        type: holding
        decode: int16
      timeout: 3s
  - name: saj.h1.grid.meter
    type: custom
    power:
      source: modbus
      id: 1
      uri: 192.168.0.20:502
      register:
        address: 0x40A1 # SmartMeterGridPowerWatt (undocumented)
        type: holding
        decode: int16
      timeout: 3s
```

h3llrais3r commented:
Small question: is there a way to also test the "battery control" options with custom meters, without creating the template?

andig commented:
Sure. See

    evcc charger --help

h3llrais3r commented:
I mean, the register works in my case with a specific firmware version.
I know that in an older firmware version, it was another register that provided the correct grid power.
That's what I mean with "too different", there is not a single register that can be used in all cases to provide the correct grid power.
So I would rather leave it out and let it up to the user to choose another meter for the grid.

h3llrais3r commented:
Ok, maybe I'm going to add it anyway, but it's the data coming from the smart meter in this case.
This gives the right data, see below:
![image](https://github.com/user-attachments/assets/8bc577c0-e07b-4658-a5b0-cc8e7781abcc)

h3llrais3r committed:
Add SAJ H1 grid meter (power and energy)

Be aware that the power is coming from the additional smart meter that should be installed together with the inverter.
Without the smart meter, the readings won't be correct or 0.

h3llrais3r commented:
Last part is the battery control, but for that I need some more time/testing and clarification about the modes, especially about the "hold" mode... I suppose this tries to keep the battery with a specific soc?
I see that it uses the `backup` mode of the inverter to keep the battery at a specific SOC, correct?
Does it always try to keep it at the `maxsoc` value or it it determined dynamically with the SOC of the battery at the time the mode is engaged?

![image](https://github.com/user-attachments/assets/21e6ab3c-3a5c-4e5b-b958-0101524a8eb0)

UDicke commented:
The "hold" just sets the backup-soc to the maxsoc value set in ui or evcc.yaml. This disables use of the battery under the set value for maxsoc and will not initiate charging,
This does only work in backup-mode, because the register for backup-soc is used for this.

UDicke commented:
![grafik](https://github.com/user-attachments/assets/b08b0c0a-f06e-48a1-9559-0e10bbc4c933)
it seems that the h1 model has a difference to h2 regarding the backup mode. Just look at the docs you provided.
this could meen, that the battery will load to a specific value when in backup mode.
Documentation about H1:
https://img.saj-electric.com/file/H1-3-6K-S2-15%C2%A0user%20manual%20EN-V0.0-2023041004191770.pdf
https://github.com/sgsancho/saj_h1_s2_modbus_esp32/blob/main/documentacion/SAJ_Modbus_Communication_Protocol_2020.pdf

in this case I would switch to "self-use mode" and propose to use register "to be find out" for "hold" 
The doc is not obvios which register migt be the one to use. I guess you have to find out which one is to use...
![grafik](https://github.com/user-attachments/assets/b0661c9b-a788-4faa-8bf3-cdef48c39a11)

premultiply commented:
"hold" means disable discharge of the battery.
If that is not possible battery should be disabled completly.

Whenever possible do not use any SoC or time related controls for battery control.

h3llrais3r commented:
Hmm, still some more investigation to be done then...

But just to be sure with the H2: 
- when you put it in `backup` mode and the backup soc on `maxsoc` (95), your battery is not charged automatically? Not even if your current battery soc is lower then the `maxsoc`? Because according to the documentation, it should only charge when battery soc is below the backup soc... 
![image](https://github.com/user-attachments/assets/4821de47-c5f9-4602-bd2c-fcc947b7b21d)
- if you say it's not charged automatically from grid, would it start charging if you have solar then?

> in this case I would switch to "self-use mode" and propose to use register "to be find out" for "hold"

I'm not sure there is a register in `self-use` mode that disables the discharge... the only one I can think of is the register that determines the total discharge power?

Besides that, the inverter also has a `passive` mode... but I think this completely deactivates everthing (solar & battery).
I see that the H2 also has this mode (according to the modus register documentation)
![image](https://github.com/user-attachments/assets/48777cd9-02ed-4652-a71f-6899d5e0b3c1)

UDicke commented:
In Backup Mode the Battery is charged from solar up to parameter 13894 "maxload" which is 100%, but not from grid. The Batterie is discharched to the value in 13902 backup soc or to 13892 "lowlimit soc" depending which register has the higher value.

I did know about the existense of passiv mode, but ignored this according to doc from saj.

The total discharge soc 13892 has Shops discharging also in self-use mode.
Here is a screenshot from  home assistant...

![image](https://github.com/user-attachments/assets/75cb5eef-cb14-4b51-b3ba-869f0e84281d)

h3llrais3r commented:
Behavior seems to be different then.

I'm having similar registers:
- 3250H #BatDisDepth (80%) -> I think this one is replaced by the 2 below where you can define lower and upper boundaries
- 3273H #BatSetSOC_H (100%)
- 3274H #BatSetSOC_L (20%)
- 3271H #BackModSOCRetain (80%) -> should correspond to the BatSocLimitKeep

H2 registers:
- 13892 = 3644h #BatOnGridDisDepth
- 13894 = 3646h #BatcharDepth
- 13905 = 3651h #BatSocLimitkeep

But apparently the behavior of the backup mode is different then for the H1, as it always charges until the backup SOC, even from grid (while for the H2 it's only with solar according to your findings)

andig commented:
> But apparently the behavior of the backup mode is different then for the H1, as it always charges until the backup SOC, even from grid (while for the H2 it's only with solar according to your findings).

Sounds like you can use `limitsoc` instead of `batterymode`?

h3llrais3r commented:
I did some digging in the code, and I think I start to understand the usage of the `limitsoc`
![image](https://github.com/user-attachments/assets/20b5fccf-0a83-4e0e-82cd-09dd4dab0fe0)

It's the register to which a soc value is written, based on the logic in the image above, right?

However, since I need to change the `appMode` of the inverter itself also to be able to "charge from grid", I probably need to write some custom logic for the `batterymode` instead, like for the H2 inverter?
So I'm coming back to my previous question if I can use the `.soc` value in the template?
![image](https://github.com/user-attachments/assets/97144f0e-b658-4a5d-8f44-897ad9d60e3b)

h3llrais3r committed:
Add configurable timeout for modbus sensors

h3llrais3r commented:
Ok... I tried to test this via a custom meters and it seems to work. (tested via `evcc meter -b <mode>`)
![image](https://github.com/user-attachments/assets/b478f06a-a23f-4e5d-9eae-4b81b338f743)

However, I would also try to test this with the template... just to be sure that the template is fine.
As far as I understood, this is only possible when running it from the devcontainer, right?

1 more question for the template: is it possible to skip a step, depending on the value of {{ .defaultmode }}?
The step in the red block is actually only required when the `defaultmode` is set to `3` (backup)
![image](https://github.com/user-attachments/assets/d0c40150-8187-44fa-8b6f-52269910eeee)
If it's not possible, no big deal to always write it, as the value is only used when the app mode is set to `3` (backup)

UDicke commented:
I am interested in your complete custom configuration. Would you please provide this?

h3llrais3r commented:
Sure, will commit it already, so you can have a look.
But I still need to properly test it, as I don't know when those "batterymode' are being set?
Is there some documentation when these modes are enabled in the flow?
Or are they all manual triggers?

h3llrais3r committed:
Add battery control (Work in progress)

h3llrais3r commented:
Hi, I finally managed to setup my dev container to run it from source with the newly created template.
However, when trying to use the battery meter, I'm getting the following error:
`[main  ] FATAL 2025/01/20 19:16:51 cannot create meter 'battery': cannot create meter type 'template': cannot create meter type 'custom': battery mode: strconv.ParseInt: parsing "<no value>": invalid syntax`

Any idea what could be wrong with my template?

h3llrais3r committed:
Fix some register types

andig commented:
> So is there a way to write the current {{ .soc }} value to another register?

Template is evaluated at startup. `{{ .soc }}` does not give you the soc at runtime! To do that at runtime you can use the `go` plugin to copy from one `in`put to one `out`put. Check the other templates, there are examples of doing so.

andig commented:
Exactly!

h3llrais3r committed:
Write current soc value to backup soc value in battery hold mode

h3llrais3r committed:
Only write backup soc when backup mode is used for battery control

h3llrais3r commented:
I'm still struggling to get this properly tested.
When running it as a custom meter, it works, but when running it via the template in the dev container, I'm getting modbus errors...
I've defined a default timeout of 5s to all modbus sources... I've done this because I sometimes get timeout error when reading an address when I'm not setting a timeout of at least 3s.  Should I define also a timeout for writing a register? (I did it in the template, but not sure if this could also cause some problem)

Can someone do a review on my template to see if I did something wrong?

andig commented:
What errors?

h3llrais3r commented:
I'll have retest to get the exact modbus errors, as I've already closed my command window and they are gone... :(
The reading parts were fine,  but from the moment I try to write, the answer of the write was not correct (I think) and it started to fail all over the place then.

But I testing this over the AIO3 module of SAJ, which seems to be known as quite unstable....
This might have something to do with it as well, because when I'm reading/writing the modbus registers manually, it mostly works.
So I'm a bit afraid that it will be very hard for me to test, due to this...

Anyway, will retry and post the errors here. (probably this evening).

h3llrais3r commented:
> > But I testing this over the AIO3 module of SAJ, which seems to be known as quite unstable....
> 
> @h3llrais3r, one hint for you: the SAJ / AIO3 connection does not like to much connections, I believe there is a limit of 2 concurrently. I use therefore [https://github.com/TCzerny/ha-modbusproxy](url) first connection for a wallbox, second for SAJ / AIO3 and third is still available. may be you did know that already or this is not the problem...

Thanks for the tip. I'm going to have a look at it. But I have some timeout from time to time with only 1 connection as well (my ha integration that reads out the data from the inverter).
It might be that I'm getting errors on the moment that my ha integration is also reading out some values (which it does every 10s). So a modbus proxy might be an idea, but I'll need to have a look at it.
I'm currently testing the saj h1 template with custom sensors, which seems to work fine at first glance (I'm using my home assistant sensors instead for reading/writing to the registers instead of doing the modbus calls directly from the template)

h3llrais3r committed:
Simplify logic

- use 'self-use' inverter mode for 'normal' battery mode
- use 'passive' inverter mode for 'hold' battery mode
- use 'time-of-use' inverter mode for 'charge' battery mode

h3llrais3r commented:
> What errors?

I took some time to test it again with the modbus over tcp, and getting the following error when making too much calls too fast:
`read failed: modbus: exception '2' (illegal data address), function '3'`

I still feel like the issue is the modbus over the tcp via the AIO3 module.
I've read on ha forum that other people are also facing similar issues with it (not being very stable) and the solution would be using a dedicated devices (like the elfin ee10a, see https://community.home-assistant.io/t/saj-solar-inverter/111932/277).

So I know the registers are correct but the readout can be unstable when using it directly over the AIO3 module.
![image](https://github.com/user-attachments/assets/82e29544-ecd1-4581-8cce-436442ed9185)

In the same home assistant community forum, they are also saying that some registers are not even writable via the AIO3 module.
This could be right, as I was unable to test the writing (also getting modbus exceptions).

So any suggestion on how to proceed with this?
Maybe we can just use the template for the readouts and skip the battery control (which needs to write to the registers)?

Or can we just keep it (but not really tested for now)?
I know the registers for battery control are also correct, as I'm currently using ha sensors for doing the writing of those registers (which work flawlessly, as it uses a far more stable connection via modbus over mqqt)

h3llrais3r commented:
Not stale.

andig commented:
> I still feel the issue is related to the modbus over tcp via the AIO3 module.

What feedback are you waiting for? Seems this is device side, not evcc?

andig commented:
Happy to continue testing, but it seems there's nothing the core team could contribute here.

h3llrais3r commented:
> use I couldn't successfully test the battery mode controls, I would remove this, and stick to the reado

I understand as you can only test if you have such an inverter.

So based on what do you accept and merge this PR?
The readouts are working fine: 
![image](https://github.com/user-attachments/assets/20227447-a66f-4346-828f-fe306a642a6d)

It's mainly the battery control I wasn't able to test properly, as I don't have an elfin device which is a far more stable device to test this.

So can I suggest to comment out the battery control for now, so that someone who want to test this in future, can at least use this template for reading out the data?

UDicke commented:
@h3llrais3r I can offer to borrow a tested AIO3 module (did work at my site up to an upgrade which included a new one).
Send me an [e-mail](mailto:u.dicke@gmx.de) and I'll send you details.
Did you connect via LAN or WLAN?

h3llrais3r commented:
> @h3llrais3r I can offer to borrow a tested AIO3 module (did work at my site up to an upgrade which included a new one). Send me an [e-mail](mailto:u.dicke@gmx.de) and I'll send you details. Did you connect via LAN or WLAN?

Replied via email. ;)

h3llrais3r commented:
Hi, I took some time to setup the devcontainer and run it from source.
Here proof that the SAJ H1 template is working by setting it up via the UI:
Setup grid meter:
![image](https://github.com/user-attachments/assets/297a3d5d-c17d-4148-a8bd-3a2a4f97267d)
Setup solar meter:
![image](https://github.com/user-attachments/assets/c127cd49-3f73-45be-9863-90bf89a1ee0f)
Setup battery:
![image](https://github.com/user-attachments/assets/5cc455b9-94f9-42b3-98d0-a60d533ceb47)
After setup:
![image](https://github.com/user-attachments/assets/f4412801-c530-4893-9728-dcdf4ab0dda4)
![image](https://github.com/user-attachments/assets/4b9663c9-3a64-44ba-97c1-3da72de2dc01)
![image](https://github.com/user-attachments/assets/5ee1073f-dde3-453b-b62a-2276818d1509)

It would be great if you could reopen this, so I can push my final changes to it to get it merged.
@andig , what do you think?

andig commented:
Seems it can‘t be reopened :(. You can just start a new one!

UDicke commented:
>  You can just start a new one!

@h3llrais3r you should just do this! 😊

h3llrais3r commented:
See https://github.com/evcc-io/evcc/pull/18843.
